### ISSUE TITLE
TryCompile raises TypeError when CacheDir is defined

### ISSUE THREAD

**Issue by pavoljuhas** [SCons/scons#3212]
I found the following problem with scons 3.0.1 when run with Python 3.7.
The behavior is the same also with the latest scons sources at e342d031c2a80cc7eb27ca139b7adee2d37a8216.
Given an otherwise empty directory with following [SConstruct](https://github.com/SCons/scons/files/2462290/SConstruct.txt) 

```python
env = Environment()
env.CacheDir('/tmp')

def MyCheck(context):
    context.Message('Checking if TryCompile works...')
    rv = context.TryCompile('int a;', '.c')
    context.Result(rv)
    return rv

conf = Configure(env, custom_tests = {'MyCheck' : MyCheck})
conf.MyCheck()
env = conf.Finish()
```

The scons run with Python 3 produces following output:

```sh
$ scons 
scons: Reading SConscript files ...
Checking if TryCompile works...TypeError : sequence item 0: expected str instance, bytes found
```

The `TypeError` is likely due to mixing of `str` and `bytes` objects in some string operation.
The problem disappears when I remove the `CacheDir` line, and the script then works as expected.

I found this on Mac OS X, but the issue is likely platform independent.
The problem does not occur when using Python 2.7.

**Comment by bdbaddog** [comment]
Please bring issues to SCons users mailing list before filing per as the bug filing template suggests.
It's also stated here.

https://scons.org/bugs.html

Likely this is another instance of a known issue with py3 and Value() nodes (which are used by Try*)

**Comment by bdbaddog** [comment]
This may only have to do with using CacheDir and python3.
I believe I have a fix, but need to make sure we create tests to cover the cause.

Try adding the following line in Node/FS.py around line 3395
```
         sigs.append(self.get_internal_path())
+
+        # Ensure that all signatures are strings before merging them to generate cachedir sig
+        sigs = [SCons.Util.to_str(s) for s in sigs]
+
         # Merge this all into a single signature
```

**Comment by pavoljuhas** [comment]
Yes, the extra line in Node/FS.py makes it work for me.  Thank you.

**Comment by bdbaddog** [comment]
I found a more appropriate place to change.
That treated the symptoms, but not the cause.

See SCons/Node/Python.py in this pull request.
https://github.com/SCons/scons/pull/3213/files

**Comment by pavoljuhas** [comment]
> See SCons/Node/Python.py in this pull request.
> https://github.com/SCons/scons/pull/3213/files

That version works as well.

**Comment by bdbaddog** [comment]
Fixed with #3213
